Data/Aeson/Types/Class.hs

```diff
@@ -250,6 +257,15 @@ class KeyValue kv where
     (.=) :: ToJSON v => Text -> v -> kv
     infixr 8 .=
 
+data ToJSONKeyFunction a
+    = ToJSONKeyText (a -> Text, a -> Encoding)
+    | ToJSONKeyValue (a -> Value, a -> Encoding)
```

Comment:
`ToJSONKeyFunction` can be [`Contravariant`](http://hackage.haskell.org/package/contravariant-1.4/docs/Data-Functor-Contravariant.html#t:Contravariant), IMHO that is quite light dependency, but might make writing instances handy (vs. providing `contramap` with new name specialised for `ToJSONKeyFunction`)